### PR TITLE
Android: Fix build on NDK r16b2

### PR DESCRIPTION
**Pull request by gwicks**
By static linking iconv (since it's missing / broken on the new unified headers) and doing more thorough checks for the `<optional>` and `<variant>` headers since they are broken under C++14 in Android

### CODE REVIEW

**Comment by JosJuice** [comment]
Simply re-adding c++1z should re-introduce the issue described in PR #6089. Could we change this PR so that we only use c++1z if we're building for Android?

File: CMakeLists.txt [hunk 1]
```diff
@@ -712,7 +712,7 @@ endif()
 find_library(ICONV_LIBRARIES NAMES iconv libiconv libiconv-2 c)
 find_path(ICONV_INCLUDE_DIR NAMES iconv.h)
 
-if (ICONV_LIBRARIES AND ICONV_INCLUDE_DIR)
+if (NOT ANDROID AND ICONV_LIBRARIES AND ICONV_INCLUDE_DIR)
```
Comment (degasus) [inline CMakeLists.txt]: I'd suggest to only execute the find_library and find_path above if not on android

File: Source/CMakeLists.txt [hunk 2]
```diff
@@ -20,9 +20,12 @@ else()
   # CMAKE_CXX_STANDARD cannot be used here because we require C++14 or newer, not any standard.
   check_and_add_flag(CXX17 -std=c++17)
   if(NOT FLAG_CXX_CXX17)
-    set(CMAKE_CXX_STANDARD 14)
-    set(CMAKE_CXX_STANDARD_REQUIRED ON)
-    set(CMAKE_CXX_EXTENSIONS OFF)
+  	check_and_add_flag(CXX1Z -std=c++1z)
```
Comment (degasus) [inline Source/CMakeLists.txt]: Is it possible to implement this logic only for android?

Comment (gwicks) [reply Source/CMakeLists.txt]: I'm not sure what the cleanest way of doing that in the CMake file would be, but I'll give it a shot

**Comment by BhaaLseN** [comment]
Any thoughts on leaving this also in the CMake file (as a comment) so others know _why_ the c++1z check is still in there? Looking it up in the history is one thing, but getting a right-in-your-face comment makes you think twice about removing a supposedly superfluous flag, as well as a more clear reason to remove it once they have fixed it.

File: Source/CMakeLists.txt [hunk 3]
```diff
@@ -18,11 +18,21 @@ if(CMAKE_C_COMPILER_ID MATCHES "MSVC")
 else()
   # Enable C++17, but fall back to C++14 if it isn't available.
   # CMAKE_CXX_STANDARD cannot be used here because we require C++14 or newer, not any standard.
-  check_and_add_flag(CXX17 -std=c++17)
-  if(NOT FLAG_CXX_CXX17)
-    set(CMAKE_CXX_STANDARD 14)
-    set(CMAKE_CXX_STANDARD_REQUIRED ON)
-    set(CMAKE_CXX_EXTENSIONS OFF)
+  # On Android, it is necessary to use C++1z instead, since the C++17 support is currently incomplete
+  if (NOT ANDROID)
```
Comment (lioncash) [inline Source/CMakeLists.txt]: This should probably be inverted (Android case coming first), so what the comment is describing is immediately below it

**Comment by JosJuice** [comment]
@dolphin-emu-bot rebuild

File: Source/CMakeLists.txt [hunk 4]
```diff
@@ -18,12 +18,12 @@ if(CMAKE_C_COMPILER_ID MATCHES "MSVC")
 else()
   # Enable C++17, but fall back to C++14 if it isn't available.
   # CMAKE_CXX_STANDARD cannot be used here because we require C++14 or newer, not any standard.
-  check_and_add_flag(CXX17 -std=c++17)
-  if(NOT FLAG_CXX_CXX17)
-    set(CMAKE_CXX_STANDARD 14)
-    set(CMAKE_CXX_STANDARD_REQUIRED ON)
-    set(CMAKE_CXX_EXTENSIONS OFF)
-  endif()
+	check_and_add_flag(CXX17 -std=c++17)
```
Comment (degasus) [inline Source/CMakeLists.txt]: unrelated whitespace change now.

**Comment by JosJuice** [comment]
@dolphin-emu-bot rebuild

**Comment by JosJuice** [comment]
@dolphin-emu-bot rebuild
